tests/web/test_lineage.py

```diff
@@ -0,0 +1,438 @@
+    assert response_json['"bar"']["col"]["models"] == {'"my_cte"': ["col"]}
+    assert response_json['"my_cte"']["col"]["models"] == {'"baz"': ["col"], '"qwe"': ["col"]}
+    assert response_json['"baz"']["col"]["models"] == {'"external_baz"': ["col"]}
+    assert response_json['"qwe"']["col"]["models"] == {'"external_qwe"': ["col"]}
+
+
+def test_get_lineage_cte_downstream_union_downstream(project_context: Context) -> None:
+    project_tmp_path = project_context.path
+    models_dir = project_tmp_path / "models"
+    models_dir.mkdir()
+    foo_sql_file = models_dir / "foo.sql"
+    foo_sql_file.write_text(
+        """MODEL (name foo);
+           SELECT col FROM bar;"""
+    )
+    bar_sql_file = models_dir / "bar.sql"
+    bar_sql_file.write_text(
+        """MODEL (name bar);
+           WITH my_cte AS (
+               SELECT * FROM baz
+           )
+           SELECT col FROM my_cte;"""
+    )
+    baz_sql_file = models_dir / "baz.sql"
+    baz_sql_file.write_text(
+        """MODEL (name baz);
+           SELECT col FROM external_table1
+           UNION
+           SELECT col FROM external_table2;"""
+    )
+    project_context.load()
+
+    response = client.get("/api/lineage/foo/col")
+    assert response.status_code == 200, response.json()
+    response_json = response.json()
+    assert response_json['"foo"']["col"]["models"] == {'"bar"': ["col"]}
+    assert response_json['"bar"']["col"]["models"] == {'"my_cte"': ["col"]}
+    assert response_json['"my_cte"']["col"]["models"] == {'"baz"': ["col"]}
+    assert response_json['"baz"']["col"]["models"] == {
+        '"external_table1"': ["col"],
+        '"external_table2"': ["col"],
+    }
+
+
+def test_get_lineage_nested_cte_union_downstream(project_context: Context) -> None:
+    project_tmp_path = project_context.path
+    models_dir = project_tmp_path / "models"
+    models_dir.mkdir()
+    foo_sql_file = models_dir / "foo.sql"
+    foo_sql_file.write_text(
+        """MODEL (name foo);
+           WITH my_cte2 AS (
+               SELECT * FROM bar
+           ), my_cte1 AS (
+               SELECT col FROM my_cte2
+               UNION
+               SELECT col FROM external_table1
+           )
+           SELECT col FROM my_cte1;"""
+    )
+    bar_sql_file = models_dir / "bar.sql"
+    bar_sql_file.write_text(
+        """MODEL (name bar);
+           SELECT col FROM external_table2
+           UNION
+           SELECT col FROM external_table3
+           ;"""
+    )
+    project_context.load()
+
+    response = client.get("/api/lineage/foo/col")
+    assert response.status_code == 200, response.json()
+    response_json = response.json()
+    assert response_json['"foo"']["col"]["models"] == {'"my_cte1"': ["col"]}
+    assert response_json['"my_cte1"']["col"]["models"] == {
+        '"my_cte2"': ["col"],
+        '"external_table1"': ["col"],
+    }
+    assert response_json['"my_cte2"']["col"]["models"] == {'"bar"': ["col"]}
```

Comment:
wouldn't it be easier just to check the whole json response as opposed to manually checking each snippet

Reply:
The whole json response can get quite cluttered with the expression and source sql.